File: iteration2/Student.java
package iteration2;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
//import java.util.Scanner;

//Inherits from the Person class and implements the ILogin interface
public class Student extends Person implements IDisplayMenu {

    //All attributes
    private Advisor advisor;
    private Transcript transcript;
    private byte gradeLevel;
    private boolean hasRequest;
    private List<Course> draft;
    private List<Course> availableCourses;

    //Implement Constructor
    public Student(int studentID, String name, String surname, String userName, String password, byte gradeLevel) {
        super(studentID, name, surname, userName, password);
        this.draft = new ArrayList<>();
        this.availableCourses = new ArrayList<>();
        this.gradeLevel = gradeLevel;
    }

    //Check if there is already a request awaiting approval
    //Check if the draft is empty
    /* Create a new Registration object and initialize it with the current
    student object (this) and draft information */
    public void sendRequest() {
        ConsoleColours.paintYellowMenu();
        if (hasRequest) {
            System.out.println("You already have a request waiting for approval.");
            ConsoleColours.resetColour();
            System.out.println();
        } else if (draft.isEmpty()) {
            System.out.println("Your draft is empty!");
            ConsoleColours.resetColour();
            System.out.println();
        } else {
            Registration registration = new Registration(this, draft);
            registration.addRequest(advisor);
        }
    }

    //Implements the printMenu method from the IDisplayMenu interface.
    //Prints a menu based on the specified menu type.
    /* After determining whether the user is a student or an advisor
    based on the String value, print the relevant menu. */
    @Override
    public void printMenu(String menuType) {
        switch (menuType) {
            case "studentMenu":
                ConsoleColours.paintBlueMenu();
                System.out.println("Welcome " + this.getName() + " " + this.getSurname() + "!");
                System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n");
                System.out.println("Please select from the following options:");
                System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
                System.out.println("                         Exit System -> 0");
                System.out.println("               Course Selection Menu -> 1");
                System.out.println("                     View Transcript -> 2");
                System.out.println("                             Log out -> 3");
                System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
                ConsoleColours.paintGreenMenu();
                System.out.print("Enter your choice: \n");
                break;
            case "courseSelectionMenu":
                ConsoleColours.paintBlueMenu();
                System.out.println("Course Selection Menu");
                System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n");
                System.out.println("Please select from the following options:");
                System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
                System.out.println("                   Back to Main Menu -> 0");
                System.out.println("                 Course Status Check -> 1");
                System.out.println("                 Add Course to Draft -> 2");
                System.out.println("            Delete Course from Draft -> 3");
                System.out.println("                 Show request status -> 4");
                System.out.println("                        Send Request -> 5");
                System.out.println("                             Log out -> 6");
                System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
                ConsoleColours.paintGreenMenu();
                System.out.print("Enter your choice: \n");
        }
    }

    // The 'addCourse' function displays all accessible courses and adds these courses to
    // the student's draft, which is a private attribute.
    public void addCourse() {
        CourseRegistrationSystem controller = new CourseRegistrationSystem();

        //If a request has been sent, terminate the function.
        if (hasRequest) {
            ConsoleColours.paintRedMenu();
            System.out.println("You can not add lecture because you have a request waiting for approval.");
            return;
        }
        //Call the calculateNumberOfCourses() function to limit with the maximum number of courses.
        int numberOfCourses = calculateNumberOfCourses();
        if (numberOfCourses >= this.getDepartment().getMaxCourseNumber()) {
            ConsoleColours.paintRedMenu();
            System.out.println("You can not add more lectures.");
            return;
        }

        //Call the viewAvailableCourses() function to see all available courses.
        ConsoleColours.paintBlueMenu();
        System.out.println("Add Course Menu");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n");
        viewAvailableCourses();

        System.out.println("      Back -> 0");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        System.out.println("Here is the available courses:");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        ConsoleColours.paintPurpleMenu();

        //Print Available Courses
        for (int i = 0; i < this.getAvailableCourses().size(); i++) {
            System.out.println((i + 1) + ". " + this.getAvailableCourses().get(i).getCourseCode() +
                    " - " + this.getAvailableCourses().get(i).getCourseName());
            System.out.println("´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´");
        }


        //Take input from the user and make assignments to the courses.
        ConsoleColours.paintGreenMenu();
        System.out.print("Choose number between 1 to " + this.getAvailableCourses().size() + " to add course: \n");

        int userNumberInput = controller.getInput();
        //Back to Course Section Menu with return value

            //If incorrect input is entered, ask for input again.
        if (1 <= userNumberInput && userNumberInput <= this.getAvailableCourses().size()) {
            this.draft.add(this.getAvailableCourses().get(userNumberInput - 1));
            this.getAvailableCourses().remove(userNumberInput - 1);
            addCourse();
        } else if (userNumberInput < 0 || userNumberInput > this.getAvailableCourses().size()){
            ConsoleColours.paintRedMenu();
            System.out.println("Invalid input, please enter a valid number!");
            addCourse();
        }
    }

    // This method calculates the total number of courses, considering both the drafted courses
    // and the courses in the transcript for which the student has not received a grade yet.
    public int calculateNumberOfCourses() {
        int numberOfCourses = draft.size();
        for (Course course : this.getTranscript().getCourseGradeMap().keySet()) {
            if (this.getTranscript().getCourseGradeMap().get(course) == null) {
                numberOfCourses++;
            }
        }
        return numberOfCourses;
    }

    public void dropCourse() {
        CourseRegistrationSystem controller = new CourseRegistrationSystem();

        // Check if the student has already sent a request
        if (hasRequest) {
            ConsoleColours.paintRedMenu();
            System.out.println("You can not remove lecture because you have a request waiting for approval.");
            return;
        }
        // Check if the draft is empty.
        if (draft.isEmpty()) {
            ConsoleColours.paintYellowMenu();
            System.out.println("Your draft is empty!");
            ConsoleColours.resetColour();
            System.out.println();
        } else {
            ConsoleColours.paintBlueMenu();
            System.out.println("Remove Course from Draft");
            System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n");
            System.out.println("               Back -> 0");
            System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
            System.out.println("Select the course you want to remove:");
            System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
            ConsoleColours.paintPurpleMenu();

            // Manipulation for displaying and removing courses in the menu.
            for (int i = 0; i < draft.size(); i++) {
                System.out.println((i + 1) + ". " + draft.get(i).getCourseCode() +
                        " - " + draft.get(i).getCourseName());
                System.out.println("´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´");
            }
            // Remove input index element -1 from the draft list if the input is valid and in the appropriate range.
            ConsoleColours.paintGreenMenu();
            System.out.print("Choose number between 1 to " + this.draft.size() + " to remove course: \n");

            int userNumberInput = controller.getInput();

            if (userNumberInput <= draft.size() && userNumberInput >= 1){
                draft.remove(userNumberInput - 1);
                dropCourse();
            }
            else if(userNumberInput > draft.size() || userNumberInput < 0) {
                ConsoleColours.paintRedMenu();
                System.out.println("Invalid input, please enter a number");
                dropCourse();
            }
        }
    }

    //Checks if the student has a registration request waiting for advisor approval
    //and prints an appropriate message to the console.
    public void showRequestStatus() {
        ConsoleColours.paintYellowMenu();
        if (hasRequest) {
            System.out.println("Your request is waiting for advisor approval.");
        } else {
            System.out.println("There is no waiting request!");
            ConsoleColours.resetColour();
            System.out.println();
            return;
        }

        ConsoleColours.paintBlueMenu();
        System.out.println("Your draft: ");
        ConsoleColours.paintPurpleMenu();
        for (Course course : draft){
            System.out.println(course.getCourseCode() + "-" + course.getCourseName());
        }
        ConsoleColours.resetColour();
        System.out.println();
    }

    public List<Course> getAvailableCourses() {
        return availableCourses;
    }

    public void viewAvailableCourses() {
        /*
        Collect all course sections from the department, and for each course, store them in
        a map<course,grade> along with the grade value. Also, keep track of course sections
        with their respective courses.
        */
        ArrayList<Course> availableCourses = new ArrayList<>();
        List<CourseSection> allCourseSections = this.getDepartment().getCourseSections();
        Map<Course, Grade> mapGrade = this.getTranscript().getCourseGradeMap();
        Map<CourseSection, Course> courseSectionCourse = this.getDepartment().getSectionCourseMap();

        /*
         While iterating through all course sections, check if the grade level is sufficient
         and if the course is already in the draft. If it is, continue the loop to change the course section.
         */
        for (CourseSection courseSection : allCourseSections) {

            if (this.gradeLevel < courseSection.getGradeLevel()
                    || draft.contains(courseSectionCourse.get(courseSection)))
                continue;

        /*
        If the conditions are not problematic and there is no course in the transcript,
        enter the 'if' statement. If status is true, then the student can add.
        If status is false, the student is ineligible unless they have taken the course before
        and received a low success grade; in this case, they can retake the course.
        */
            Course course = courseSectionCourse.get(courseSection);

            if (!mapGrade.containsKey(course)) {
                boolean status = true;
                for (Course prerequisite : course.getPreRequisiteCourses()) {
                    if ((mapGrade.get(prerequisite) == null
                            || mapGrade.get(prerequisite).getLetterGrade().equals("FF")
                            || mapGrade.get(prerequisite).getLetterGrade().equals("FD"))) {
                        status = false;
                    }
                }
                if (status)
                    availableCourses.add(course);
            } else {
                // If the student has previously taken the course and has a grade lower than 'CC', include it
                if (mapGrade.get(course) != null && mapGrade.get(course).getLetterGrade().compareTo("CC") > 0) {
                    availableCourses.add(course);
                }
            }
        }
        this.availableCourses = availableCourses;
    }

    //Checks if the provided username and password match the student's credentials.
    //Overrides the abstract login method in the Person class.
    @Override
    public boolean login(String userName, String password) {
        return this.getUserName().equals(userName) && this.getPassword().equals(password);
    }

    public String getDepartmentName() {
        return this.getDepartment().getDepartmentName();
    }

    public void setPassword(String password) {
        super.setPassword(password);
    }

    // Override the setUserName method to set the username by
    // calling the setUserName method from the inherited Person class.
    @Override
    public void setUserName(String userName) {
        super.setUserName(userName);
    }

    public Advisor getAdvisor() {
        return advisor;
    }

    public void setAdvisor(Advisor advisor) {
        this.advisor = advisor;
    }

    public byte getGradeLevel() {
        return gradeLevel;
    }

    public void setHasRequest(boolean hasRequest) {
        this.hasRequest = hasRequest;
    }

    public List<Course> getDraft() {
        return draft;
    }

    public Transcript getTranscript() {
        return transcript;
    }

    public void setTranscript(Transcript transcript) {
        this.transcript = transcript;
    }

    public void setDraft(List<Course> draft) {
        this.draft = draft;
    }

    public boolean isHasRequest() {
        return hasRequest;
    }

}

File: iteration3/Student.py
import logging
from iteration3.ConsoleColours import ConsoleColours
from iteration3.Course import Course
from iteration3.IDisplayMenu import IDisplayMenu
from iteration3.Person import Person


class Student(Person, IDisplayMenu):
    def __init__(self, studentID, name, surname, userName, password, semester):
        super().__init__(studentID, name, surname, userName, password)
        from iteration3.CourseRegistrationSystem import CourseRegistrationSystem
        self.__system = CourseRegistrationSystem()
        self.__draft = []
        self.__availableCoursesToAdd = []
        self.__availableCoursesToDrop = []
        self.__labSections = []
        self.__hasRequest = False
        self.__transcript = None
        self.__advisor = None
        self.__notification = None
        self.__semester = semester
        self.__schedule = [[None for _ in range(8)] for _ in range(5)]

    def sendRequest(self):
        """
        Sends a request for approval to the assigned advisor.

        If the student already has a request waiting for approval, it prints a warning message and does nothing.
        If the student's draft is empty, it prints a message and does nothing.
        Otherwise, it creates a new Registration object with the student and draft as parameters, adds the advisor to the request,
        and logs the event.

        Returns:
            None
        """
        ConsoleColours.paintRedMenu()
        if self.__hasRequest:
            print("You already have a request waiting for approval.")
            logging.warning(
                f"Student {self.getID()} already has a request waiting for {self.__advisor.getID()}'s approval.")
            ConsoleColours.resetColour()
        elif not self.__draft:
            print("Your draft is empty!")
            ConsoleColours.resetColour()
        else:
            from iteration3.Registration import Registration
            registration = Registration(self, self.__draft)
            registration.addRequest(self.__advisor)
            logging.info(f"Student {self.getID()} sent a request to advisor {self.__advisor.getID()}.")

    def printMenu(self, menuType):
        """
        Prints out the menu based on the given menuType.

        Parameters:
            menuType (str): The type of menu to be printed.

        Returns:
            None
        """
        match menuType:
            case "studentMenu":
                ConsoleColours.paintBlueMenu()
                print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
                print(f"Welcome {self.getName()} {self.getSurname()}!")
                print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n")

                if self.__notification is not None:
                    ConsoleColours.paintYellowMenu()
                    print(self.__notification.message)
                    self.__notification = None
                    ConsoleColours.paintBlueMenu()

                print("Please select from the following options:")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                print("                         Exit System -> 0")
                print("                 Course Registration -> 1")
                print("                     View Transcript -> 2")
                print("                             Log out -> 3")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                ConsoleColours.paintGreenMenu()
                print("Enter your choice: ")

            case "courseRegistrationMenu":
                ConsoleColours.paintBlueMenu()
                print("Course Registration Menu")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n")
                print("Please select from the following options:")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                print("                   Back to Main Menu -> 0")
                print("                 Course Status Check -> 1")
                print("                 Add Course to Draft -> 2")
                print("            Delete Course from Draft -> 3")
                print("                 Show request status -> 4")
                print("                        Send Request -> 5")
                print("                             Log out -> 6")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                ConsoleColours.paintGreenMenu()
                print("Enter your choice: ")

            case "courseSelectionMenu":
                ConsoleColours.paintBlueMenu()
                print("Course Selection Menu")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n")
                print("Please select from the following options:")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                print("    Back to Course Registration Menu -> 0")
                print("                       Add Mandatory -> 1")
                print("              Add Technical Elective -> 2")
                print("          Add Non-Technical Elective -> 3")
                print("      Add Faculty Technical Elective -> 4")
                print("                         Drop Course -> 5")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                ConsoleColours.paintGreenMenu()
                print("Enter your choice: ")
            case "addCourse":
                ConsoleColours.paintBlueMenu()
                print("Add Course Menu")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n")
                print("      Back -> 0")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                print("Here is the available courses to add:")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                ConsoleColours.paintPurpleMenu()

            case "dropCourse":
                ConsoleColours.paintBlueMenu()
                print("Drop Course Menu")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n")
                print("       Back -> 0")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                print("Here is the available courses to drop:")
                print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
                ConsoleColours.paintPurpleMenu()

    def addCourseToDraft(self):
        try:
            if self.__hasRequest:
                ConsoleColours.paintRedMenu()
                print("You can not add lecture because you have a request waiting for approval.")
                logging.warning(
                    f"Student {self.getID()} can not add lecture because he/she has a request waiting for {self.__advisor.getID()}'s approval.")
                return
            ConsoleColours.paintBlueMenu()
            self.printMenu("courseSelectionMenu")
            choice = self.__system.getInput()
            match choice:
                case 0:
                    return
                case 1:
                    self.addMandatoryCourse()
                case 2:
                    self.addTechnicalElective()
                case 3:
                    self.addNonTechnicalElective()
                case 4:
                    self.addFacultyTechnicalElective()
                case 5:
                    self.addCourseToDrop()
                case -1:
                    ConsoleColours.paintRedMenu()
                    print("Please enter valid number!")
                case _:
                    ConsoleColours.paintRedMenu()
                    print("Invalid input, please enter a valid number!")
            self.addCourseToDraft()
        except Exception as e:
            logging.error(f"An error occurred while adding course to draft: {e}")
            print("Something went wrong. Please try again..")
            self.addCourseToDraft()

    def addMandatoryCourse(self):
        if self.maxCoursesReached():
            return
        self.computeAvailableMandatoryCourses()

        if not self.__availableCoursesToAdd:
            ConsoleColours.paintRedMenu()
            print("You do not have any addable course!")
        else:
            self.printMenu("addCourse")
            for i, course in enumerate(self.__availableCoursesToAdd):
                print(
                    f"{i + 1}. {course.getCourseCode()} - {course.getCourseName()} - {course.getDay()} - {course.getHour()}.00")
                print("´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´")

            ConsoleColours.paintGreenMenu()
            print(f"Choose number between 1 to {len(self.__availableCoursesToAdd)} to add course: \n")
            userNumberInput1 = self.__system.getInput()

            if 1 <= userNumberInput1 <= len(self.__availableCoursesToAdd):
                self.chooseLabSection(self.__availableCoursesToAdd[userNumberInput1 - 1])
                logging.info(
                    f"Student {self.getID()} added {self.__availableCoursesToAdd[userNumberInput1 - 1].getCourseCode()} to draft.")
                self.__draft.append(self.__availableCoursesToAdd[userNumberInput1 - 1])
                del self.__availableCoursesToAdd[userNumberInput1 - 1]
                if self.__availableCoursesToAdd:
                    self.addMandatoryCourse()
            elif userNumberInput1 > len(self.__availableCoursesToAdd) or userNumberInput1 < 0:
                ConsoleColours.paintRedMenu()
                logging.warning(f"Student {self.getID()} entered invalid input.")
                print("Invalid input, please enter a valid number")
                self.addMandatoryCourse()
            else:
                return

    def chooseLabSection(self, course):
        if not course.getLaboratorySections():
            return
        availableLabSections = []
        ConsoleColours.paintBlueMenu()
        print("Here is the available laboratory sections to add:")
        ConsoleColours.paintPurpleMenu()
        for labSection in course.getLaboratorySections():
            labCourse = Course(None, labSection.getLaboratorySectionCode(), None, 0, 0, 0, labSection.getHour(),
                               labSection.getDay())
            if self.hasCourseOverlap(labCourse, True) or labSection.getCapacity() <= labSection.getNumberOfStudents():
                continue
            else:
                availableLabSections.append(labSection)

        for i, labSection in enumerate(availableLabSections):
            print(
                f"{i + 1}. {labSection.getLaboratorySectionCode()} - {course.getCourseName()} (Lab) - {labSection.getDay()} - {labSection.getHour()}.00")
            print("´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´")
        ConsoleColours.paintGreenMenu()
        print(f"Choose number between 1 to {len(availableLabSections)} to add laboratory section: \n")
        userNumberInput = self.__system.getInput()

        if 1 <= userNumberInput <= len(availableLabSections):
            self.__labSections.append(availableLabSections[userNumberInput - 1])
            availableLabSections[userNumberInput - 1].setNumberOfStudents(
                availableLabSections[userNumberInput - 1].getNumberOfStudents() + 1)
        elif userNumberInput > len(availableLabSections) or userNumberInput <= 0:
            ConsoleColours.paintRedMenu()
            print("Invalid input, please enter a valid number")
            ConsoleColours.resetColour()
            self.chooseLabSection(course)
        else:
            return

    def addTechnicalElective(self):
        if self.maxCoursesReached():
            return
        self.computeAvailableTEAndFTECourses("TE")

        if not self.__availableCoursesToAdd:
            ConsoleColours.paintRedMenu()
            print("You do not have any addable course!")
        else:
            self.printMenu("addCourse")
            for i, course in enumerate(self.__availableCoursesToAdd):
                print(
                    f"{i + 1}. {course.getCourseCode()} - {course.getCourseName()} - {course.getDay()} - {course.getHour()}.00")
                print("´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´")

            ConsoleColours.paintGreenMenu()
            print(f"Choose number between 1 to {len(self.__availableCoursesToAdd)} to add course: \n")
            userNumberInput1 = self.__system.getInput()

            if 1 <= userNumberInput1 <= len(self.__availableCoursesToAdd):
                self.__draft.append(self.__availableCoursesToAdd[userNumberInput1 - 1])
                logging.info(
                    f"Student {self.getID()} added {self.__availableCoursesToAdd[userNumberInput1 - 1].getCourseCode()} to draft.")
                del self.__availableCoursesToAdd[userNumberInput1 - 1]
                if self.__availableCoursesToAdd:
                    self.addTechnicalElective()
            elif userNumberInput1 > len(self.__availableCoursesToAdd) or userNumberInput1 < 0:
                ConsoleColours.paintRedMenu()
                print("Invalid input, please enter a valid number")
                self.addTechnicalElective()
            else:
                return

    def computeAvailableTEAndFTECourses(self, courseType):
        self.__availableCoursesToAdd.clear()
        sum = 0
        if self.__transcript:
            mapGrade = self.__transcript.getCourseGradeMap()
            studentCourses = self.__transcript.getStudentCourses()
            if studentCourses:
                for course in studentCourses:
                    grades = mapGrade.get(course, [])
                    lastGrade = grades[-1] if grades else None
                    if (course.getSemester() > 6
                            or course.getCourseType() == "NTE"
                            or lastGrade is None
                            or lastGrade.getLetterGrade() is None
                            or lastGrade.getLetterGrade() > "DD"):
                        continue
                    sum += course.getCourseCredit()

        if sum >= 155:
            for course in self.getDepartment().getCourses():
                if (course.getCourseType() != courseType
                        or not course.hasCapacity()
                        or self.hasCourseOverlap(course, False)
                        or (self.__semester < course.getSemester() and self.__transcript.getCgpa() < 3)
                        or course in self.__draft
                        or not self.checkThePrerequisiteAndCourseThatWasTaken(course)
                        or (course.getSemester() % 2 != self.__semester % 2)):
                    continue
                else:
                    self.__availableCoursesToAdd.append(course)

    def addFacultyTechnicalElective(self):
        if self.maxCoursesReached():
            return
        self.computeAvailableTEAndFTECourses("FTE")

        if not self.__availableCoursesToAdd:
            ConsoleColours.paintRedMenu()
            print("You do not have any addable course!")
        else:
            self.printMenu("addCourse")
            for i, course in enumerate(self.__availableCoursesToAdd):
                print(
                    f"{i + 1}. {course.getCourseCode()} - {course.getCourseName()} - {course.getDay()} - {course.getHour()}.00")
                print("´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´´")

            ConsoleColours.paintGreenMenu()
            print(f"Choose number between 1 to {len(self.__availableCoursesToAdd)} to add course: \n")
            userNumberInput1 = self.__system.getInput()

            if 1 <= userNumberInput1 <= len(self.__availableCoursesToAdd):
                self.__draft.append(self.__availableCoursesToAdd[userNumberInput1 - 1])
                logging.info(
                    f"Student {self.getID()} added {self.__availableCoursesToAdd[userNumberInput1 - 1].getCourseCode()} to draft.")
                del self.__availableCoursesToAdd[userNumberInput1 - 1]
                if self.__availableCoursesToAdd:
                    self.addFacultyTechnicalElective()
            elif userNumberInput1 > len(self.__availableCoursesToAdd) or userNumberInput1 < 0:
                ConsoleColours.paintRedMenu()
                print("Invalid input, please enter a valid number")
                self.addFacultyTechnicalElective()
            else:
                return